dotnet e2e

IsaacCloos opened the pull request:
# Summary

Extend e2e coverage to include net6.0 runtime and Kestrel web server.

# Coverage

I've been running the sanity tests against these environments

- [X] macos arm64 (local)
- [X] linux bullseye arm64 (container)
- [ ] linux buster x86_64 (container) [coming soon]

# Tasks

- [X] Make tests
- [ ] Get Feedback on work
- [ ] Update change log after feedback and code updates
- [ ] Update CI to run tests

---

# Criticism and feedback welcome 😄 

Drafting a pr for feedback. I will have all of the tasks above completed (including squashing and changelog as the guideline dictates) before I move from the draft state. 

I have a few questions about the code myself, but that can wait till tomorrow 😴

infiniteregrets commented:
hi @IsaacCloos, thank you for the PR! I'll be reviewing this soon

IsaacCloos commented:
Absolutely no rush! I'm happy to elaborate on any of the decisions made in these tests, their format, or changes made to the sanity suite.

Furthermore, I'm happy to make any necessary revisions to make this work valuable to the project 😄

infiniteregrets commented on tests/dotnet-e2e/fileops/Program.cs:
```diff
@@ -0,0 +1,78 @@
+
+    if (dat != TEXT)
+        throw new Exception($"Expected {TEXT}, got {dat}");
+}
```
Not completly sure how dotnet works, but is there a testing framework? why not have proper function names?

IsaacCloos replied:
There are three primary testing frameworks in the dotnet ecosystem (atm).

- NUnit
- xUnit
- MsTest

There are pros and cons to using each of them. None of them are included out-of-the-box in a default console application template like I've constructed here. Adding any of them would involve more dependencies, increased project complexity, and running different commands against the project root- not in a "build this file and execute it" format like is currently the arrangement with the Go tests.

All that said, using one of these frameworks would add complexity while ultimately doing the same thing I've done here with vanilla operators and Exception throwing.

---

As for functions vs scopes, that was just a style choice 😄  I will get that changed 👍🏻

infiniteregrets replied:
sounds good, I guess we dont need a framework then

infiniteregrets commented on tests/dotnet-e2e/fileops/Program.cs:
```diff
@@ -0,0 +1,78 @@
+    // Arg 2 specifies offset for stream
+    fileStream.Read(buf, 0, buf.Length);
+
+    if (System.Text.Encoding.Default.GetString(buf) != TEXT)
+        throw new Exception($"Expected {TEXT}, got {System.Text.Encoding.Default.GetString(buf)}");
+
```
Can you add assertions instead?

IsaacCloos replied:
If you mean assertions in the sense of a dotnet equivalent to `assert_eq!` to reduce code for comparing two str and then printing the mismatched results with an Exception, then no.

 In dotnet there isn't a built-in test framework (see linked comment below for details). If you are referring to something else, let me know 😃 

https://github.com/metalbear-co/mirrord/pull/480#discussion_r987475661

infiniteregrets replied:
yeah I was referring to the same point as before, this is fine

infiniteregrets commented on tests/dotnet-e2e/fileops/Program.cs:
```diff
@@ -0,0 +1,78 @@
+    return filePath;
+}
+
+void checkFileExistsOnHost(string fileName)
+{
+    if (File.Exists(fileName))
+        throw new Exception("file exists on host");
+}
```
Once we add support for `stat`, we'd probably not need this at all. So not sure if we should check if the file exists on host

IsaacCloos replied:
Excellent! Removed 👍🏻

infiniteregrets commented on tests/dotnet-e2e/fileops/Program.cs:
```diff
@@ -0,0 +1,78 @@
+    var buf = new byte[TEXT.Length];
+
+    // https://learn.microsoft.com/en-us/dotnet/api/system.io.filestream.canseek?view=net-6.0
+    if (!fileStream.CanSeek)
+        throw new Exception("file stream CanSeek failure");
```
Do we really need this check?

IsaacCloos replied:
I was following the Go tests line-for-line. Perhaps too closely! Upon review, I don't think this check adds value to the Lseek code block in this file.

Removed 👍🏻

infiniteregrets commented on tests/dotnet-e2e/fileops/Program.cs:
```diff
@@ -0,0 +1,78 @@
+}
+
+
+// Faccessat
+// TODO: syscall limitations x-platform. Will revisit after discussion
```
What's x-platform?

IsaacCloos replied:
Whoops, I left in that short-hand for "cross platform"; in this case referring to Macos and Linux. 

The point of this comment is that I was unable to verify if the intended system calls were being made on any or all platforms the e2e/e2e-macos sections of the `ci.yml` file covers.

This Faccessat check is currently in the Python and Go test suites (which together cover both the ci sections mentioned above). This code block as it stands is probably not adding value. With your leave I can remove it for now 👍🏻

infiniteregrets replied:
sure, remove it

infiniteregrets reviewed (commented):
Awesome work! thanks for adding the tests 😄 . Left a few comments. Could you please resolve the merge conflicts?

IsaacCloos commented:
Thank you for the feedback! 

I will create a new PR tomorrow with an appropriate GIT history 😅

infiniteregrets commented:
@IsaacCloos please add steps in the CI to compile dotnet files, and run those binaries with mirrord

IsaacCloos commented:
I added the build instructions to both `e2e` and `e2e_macos`

For `e2e` I believe the new checks are triggered by this catch all.
https://github.com/metalbear-co/mirrord/blob/ee7d86403677e85b472c11a27be93b30a83c0d5b/.github/workflows/ci.yaml#L249-L250

For `e2e_macos` I added specific calls to trigger these tests:
```yml
- name: cargo test
        run: cargo test -p tests test_mirror_http_traffic
        run: cargo test -p tests test_dotnet6_outgoing_traffic_single_request_enabled
        run: cargo test -p tests test_dotnet6_remote_env_vars_works
        run: cargo test -p tests test_dotnet6_dns_lookup
```

I'd love feedback on how these have been set up here 😄

infiniteregrets commented:
> I added the build instructions to both `e2e` and `e2e_macos`
> 
> For `e2e` I believe the new checks are triggered by this catch all.
> 
> https://github.com/metalbear-co/mirrord/blob/ee7d86403677e85b472c11a27be93b30a83c0d5b/.github/workflows/ci.yaml#L249-L250
> 
> For `e2e_macos` I added specific calls to trigger these tests:
> 
> ```yaml
> - name: cargo test
>         run: |
>           cargo test -p tests test_mirror_http_traffic
>           cargo test -p tests test_dotnet6_outgoing_traffic_single_request_enabled
>           cargo test -p tests test_dotnet6_remote_env_vars_works
>           cargo test -p tests test_dotnet6_dns_lookup
> ```
> 
> I'd love feedback on how these have been set up here 😄

Awesome work so far! But I am not sure if I understand, but what's exactly wrong with using `cargo test -p tests`?

infiniteregrets commented on .github/workflows/release.yaml:
```diff
@@ -166,40 +166,18 @@ jobs:
-          npm install          
-          tsc && vsce package -o /tmp/vscode-ext/vscode-ext-linux-x64 --target linux-x64
-          vsce publish ${{ steps.version.outputs.version }} --pre-release --target linux-x64
+      - run: cd vscode-ext && npm install && tsc && vsce publish ${{ steps.version.outputs.version }} --pre-release --target linux-x64
```
Can you please merge main

infiniteregrets commented on tests/dotnet-e2e/dns/Program.cs:
```diff
@@ -0,0 +1,6 @@
+var url = "https://google.com";
+Uri myUri = new Uri(url);
+var ip = (await Dns.GetHostAddressesAsync(myUri.Host))[0];
+Console.WriteLine(ip);
```
II dont tthink we need to write to console? rather just exit the process, and do something along the lines of https://github.com/metalbear-co/mirrord/blob/main/tests/node-e2e/remote_dns/test_remote_dns_lookup_google.mjs#L19

IsaacCloos replied:
Absolutely. Let me make this change 👍🏻

infiniteregrets commented on tests/dotnet-e2e/fileops/Program.cs:
```diff
@@ -0,0 +1,50 @@
+TestWrite();
+TestLseek();
+
+void TestRead()
+{
+    var dat = File.ReadAllText("/app/test.txt");
```
seems like when these test are run there is a pipe created as seen in this log:
> stdout 2022-10-05T22:09:58.418189143Z 12402: 2022-10-05T22:09:58.417399Z DEBUG ThreadId(04) mirrord_layer::file::hooks: open_logic -> path "/tmp/clr-debug-pipe-12402-64281-in"
relevant issue: https://github.com/dotnet/runtime/issues/7204
could you please somehow disable this?

or else add the file pattern to this ignore regex https://github.com/metalbear-co/mirrord/blob/main/mirrord-layer/src/file.rs#L32

IsaacCloos replied:
Wow! Thank you for bringing this to my attention and linking the relevant issues. 

I will research this and make any necessary changes to fix this 👍🏻

infiniteregrets replied:
Thanks Issac, let us know if you need any help or feel free to reach out on [discord](https://discord.com/invite/J5YSrStDKD)

IsaacCloos replied:
I was able to resolve this with:
```
let mut env = HashMap::new();
env.insert("COMPlus_EnableDiagnostics", "0"); <--------------
env.insert("MIRRORD_AGENT_IMAGE", "test");
env.insert("MIRRORD_CHECK_VERSION", "false");
env.insert("MIRRORD_AGENT_RUST_LOG", "warn,mirrord=debug");
env.insert("MIRRORD_AGENT_COMMUNICATION_TIMEOUT", "180");
env.insert("RUST_LOG", "warn,mirrord=debug");
```

Reference:

https://github.com/dotnet/runtime/issues/7204#issuecomment-723204695

Thanks again for pointing me toward this discussion 😄

infiniteregrets replied:
Awesome, seems to have been fixed

infiniteregrets reviewed (changes requested):
Looking good

infiniteregrets commented:
@IsaacCloos do the fileops tests work locally for you? can you send the strace output the fileops? maybe we are not hooking a function that dotnet uses

IsaacCloos commented:
I'm sorry for the delay! Let me get you all of my local info when I get home 🏡 

And yes, all tests are working for me on the platforms listed above. I have been testing Linux via docker containers

infiniteregrets commented:
no worries! you mention macos arm64 in the coverage, do these work there?

IsaacCloos commented:
> no worries! you mention macos arm64 in the coverage, do these work there?

Yup! These are working on my local machine and the arm64 linux distro 'bullseye'.

I am currently rebuilding the x86_64 container (buster) again to test there too, but it's taking a long time 😅

infiniteregrets commented on tests/src/sanity.rs:
```diff
@@ -286,6 +290,7 @@ mod tests {
+        env.insert("COMPlus_EnableDiagnostics", "0");
         env.insert("MIRRORD_AGENT_IMAGE", "test");
         env.insert("MIRRORD_CHECK_VERSION", "false");
         env.insert("MIRRORD_AGENT_RUST_LOG", "warn,mirrord=debug");
```
can you please change all log levels to trace, might provide more context what is going wrong

IsaacCloos replied:
Thank you!

[Here](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-6.0#set-log-level-by-command-line-environment-variables-and-other-configuration) is dotnet specific instructions for setting the log level as well.

infiniteregrets replied:
nice, but for now can you change it back to 0 so that we can just focus on file-based logs? thanks!

IsaacCloos replied:
# 😅
Sorry! I got carried away with enabling the tracing. I'll fix this

IsaacCloos commented:
[Here](https://github.com/metalbear-co/mirrord/actions/runs/3206453151) you can find a failed execution of the `fileops` stage with tracing enabled and the dotnet diagnostics disabled.

I am a bit out of my depth comparing why these tests work locally and not in CI, but I'm still digging in as much as I can 👍🏻 

also,
I really appreciate all of the time you've spent helping me 😄 Sorry it's been so problematic

infiniteregrets commented:
@IsaacCloos seems like dotnet uses `pread`, as @meowjesty found and has implemented a hook for it, so lets wait for it to get merged and then we'll check if our tests pass

infiniteregrets commented:
hi @IsaacCloos, the pread hook has been merged, could you please resolve the conflicts?

IsaacCloos commented:
I've been following along with the related work and learning a lot about mirrord and more advanced Rust code 📓 

I'm sorry the nature of this work quickly moved into aspects of the project I wasn't ready to write confidently, so I really appreciate all the work people have been putting in to allow these tests to be successful and possibly even helpful!

With that said, I found this comment on the latest related pr:
https://github.com/metalbear-co/mirrord/pull/558#discussion_r994149087

The mirrord test suite has evolved quite a bit since I first opened this PR. Would these tests be more helpful as integration tests? Are there ways I can redo/change this work to be more valuable to the project? 

Thanks again 😄

IsaacCloos commented:
I'd be more than happy to make a new PR to convert these e2e tests to integration tests 😄  

I also will keep this draft alive until related works are merged so I can have some closure on the `pread` hook 👍🏻 

---
Also, are these outstanding tickets better suited to integration tests as well? 
> #199
#241 

I've been investigating the necessary work to possibly take a stab at #241 after #480 is satisfied 🤞🏻

aviramha commented:
@IsaacCloos Great!
Yes those are better done as integration tests as well ;)

Thanks !!

aviramha commented:
Thanks for being involved and staying in loop while we had some issues @IsaacCloos .
Agree on the great work by @infiniteregrets :)

infiniteregrets commented:
thanks for the contributions @IsaacCloos, great to see the e2e is passing! happy to help (:

since you have tested a lot, feel free to add docs at https://github.com/metalbear-co/mirrord.dev if you get time
